### PR TITLE
Samba4ad: New Fixes. 

### PR DESCRIPTION
**Pull request by abprasa**
# Description
Modified the old samba4 configuration to make it more dynamic and also fixed the frequent appearing resolution issues that we have come across during earlier testing.

# Impacts
Possible impact during backporting while replacing MS-AD with Samba4.

# Issue
1. Fixed the host's file to remove DNS resolution errors seen post-reload.
2. To have the config persistent after reloading updated the dhclient.conf file with the nameserver configured for the domain_name listed in the correct order.
3. Introduced new variables to make the code more dynamic and avoid duplicate reinstallation.

# Delete branch after the merge
YES 

# Checklist
- [ ] Document the feature

# New Features
* Doing venom test samba4 AD.

### CODE REVIEW

File: addons/vagrant/inventory/group_vars/linux_servers/samba4ad.yml [hunk 1]
```diff
@@ -1,4 +1,6 @@
 ---
+samba4ad__mgmt_ip: "{{ hostvars[inventory_hostname]['mgmt_ip'] }}"
```
Comment (nqb) [inline addons/vagrant/inventory/group_vars/linux_servers/samba4ad.yml:2]:
```suggestion
samba4ad__mgmt_ip: "{{ packetfence_server_mgmt_ip }}"
```

File: addons/vagrant/playbooks/linux_servers/samba4ad.yml [hunk 2]
```diff
@@ -9,190 +9,217 @@
...
-      src: "/etc/krb5.conf"
-      dest: "/var/lib/samba/private/krb5.conf"
-      state: link
-
-  - name: Unmask and start Services
-    ansible.builtin.systemd:
-      enabled: yes
+    check_mode: yes
+    register: check_if_samba
+    ignore_errors: true
```
Comment (nqb) [inline addons/vagrant/playbooks/linux_servers/samba4ad.yml:12]: I don't think you need "ignore_errors"

File: addons/vagrant/playbooks/linux_servers/samba4ad.yml [hunk 3]
```diff
@@ -9,190 +9,217 @@
...
+        cmd: "{{ item }}"
+      loop:
+      - "samba-tool user setexpiry {{ samba4ad__username }} --noexpiry"
+      - "samba-tool user setexpiry {{ packetfence_domain_account }} --noexpiry"
+      - "samba-tool group add sponsor"
+      - "samba-tool group add IT"
+      - "samba-tool group addmembers sponsor {{ packetfence_domain_account }}"
+      - "samba-tool group addmembers IT {{ packetfence_domain_account }}"
+
+    when: check_if_samba.failed is true
```
Comment (nqb) [inline addons/vagrant/playbooks/linux_servers/samba4ad.yml]: If `ansible.builtin.lineinfile` module returns "CHANGED" when it found a line in a file (even in check mode), you can certainly simplify this line by:

`when: check_if_samba is changed`

Which is, IMHO, more readable.

**Review by nqb — CHANGES_REQUESTED** [review]
Using the `check_if_samba` variable, you can certainly skip more tasks on an existing installation.

At the moment, you only skip "Configure Samba AD Stuff" task but there is no need to re-run previous tasks (especially reboot).

What you can do is to move "installation" tasks into a dedicated playbook called "installation.yml" and use a condition based on `check_if_samba` variable to run that playbook ([example](https://github.com/inverse-inc/packetfence/blob/1aaf6ec7e5ee3e5f688e371262af8502111c64fb/t/venom/scenarios/template/site.yml#L6-L7))

File: addons/vagrant/playbooks/linux_servers/samba4ad.yml (hunk 2 above)
Comment (JeGoi) [reply addons/vagrant/playbooks/linux_servers/samba4ad.yml:12]: Unfortunately, It does not work without it on a fresh install. I think it is due to the ansible version that we use.
It is joining the fact that our tasks order may need to be reviewed.
There are currently following the samba4ad wiki.

On a fresh installation with the current tasks order:
```
TASK [Check if samba4ad configuration is there] ********************************
Friday 21 April 2023  08:05:00 -0400 (0:00:00.827)       0:00:04.151 **********
fatal: [ad]: FAILED! => changed=false
  msg: Destination /etc/samba/smb.conf does not exist !
  rc: 257
...ignoring
```
Without the ignore, the playbook is failing.
The best should be to install samba first and them test if the file is here and modified, but I am not sure how to handle the multiple painful reboot

File: addons/vagrant/playbooks/linux_servers/samba4ad.yml (hunk 3 above)
Comment (JeGoi) [reply addons/vagrant/playbooks/linux_servers/samba4ad.yml]: This is not working when there is no installation.
To use that we need to change tasks order.

**Comment by JeGoi** [comment]
Nope, we skip almost everything due to the block line 21.
It is like, the file is here and contains the right config then just restart the service.
But you are right about the installation. It should be better but take more time.

**Pull request merged**